### PR TITLE
Add command `validate` to crm_resource

### PR DESCRIPTION
**Pull request by ioguix**
Hi,

Based on the feedback from users deploying the PAF resource agent, it is quite easy to forget a step in the resource setup often leading to resource failure or node fencing.

As the `validate-all` function of resource agent is intended to be used by UIs, it seems useful to add command to call it from a cli tool.

I have been able to check it with a simple scenario like:

~~~console
# crm_shadow --create test_validate

# [...cluster setup...]

# crm_resource -r pgsql-ha --validate --verbose
Operation validate-all for pgsqld:0 (ocf:heartbeat:pgsqlms) returned 2
 >  stderr: ocf-exit-reason:Recovery template file must contain in primary_conninfo parameter "application_name=srv1"
~~~

The next step would be to be able to run this command on remote nodes as well, but I am not sure yet how to achieve this. Is it possible to export/import the cib in the crm_shadow environment?

Regards,

### CODE REVIEW

**Comment by oalbrigt** [comment]
You could add it to the start() function of the resoure agent like in LVM (and a couple of other agents):
https://github.com/oalbrigt/resource-agents/blob/805bfa13543d2356eb4b7e72c040f829a24a3c69/heartbeat/LVM#L716

**Comment by ioguix** [comment]
@oalbrigt,

Well, we actually call it on almost all actions, like some other RA in fact (pgsql, postfix, IPaddr2, galera, mysql, ...), not just on the start one. See: https://github.com/dalibo/PAF/blob/master/script/pgsqlms#L1991

That's why after the probe action, should an error occurs during validate-all:
* the RM tries to stop the resource
* the stop action call validate-all again, raising the same error again
* the RM decide to fence the node

**Comment by kgaillot** [comment]
Thanks, this is a good addition!

e44766c is not a bad idea, but I don't want to change the existing behavior, in case anyone has come to rely on it. You can disable the output when using validate, though.

File: tools/crm_resource_runtime.c [hunk 1]
```diff
@@ -1552,6 +1552,8 @@ cli_resource_execute(const char *rsc_id, const char *rsc_action, GHashTable *ove
                    action, rsc->id, rclass, rprov ? rprov : "", rtype, op->status);
         }
 
+        if(!do_trace) goto done;
+
         if (op->stdout_data) {
```
Comment (kgaillot) [inline tools/crm_resource_runtime.c]: I'd rather not force verbose on all other commands. I think the easiest approach would be to add here:

    && (do_trace || safe_str_neq(rsc_action, "validate-all")

**Comment by ioguix** [comment]
Hi @kgaillot,

I fixed the code the way you suggested, but merged it in the previous commit and force push to the branch.

File: tools/crm_resource_runtime.c [hunk 2]
```diff
@@ -1552,6 +1552,10 @@ cli_resource_execute(const char *rsc_id, const char *rsc_action, GHashTable *ove
                    action, rsc->id, rclass, rprov ? rprov : "", rtype, op->status);
         }
 
+        /* hide output for validate-all if not in verbose */
+        if (!do_trace && safe_str_neq(rsc_action, "validate-all"))
```
Comment (kgaillot) [inline tools/crm_resource_runtime.c]: I think we need safe_str_eq() here

**Comment by ioguix** [comment]
What a shame... Copied/pasted/changed your code snippet without paying attention to the function name :(
It's fixed now, a tested :)

**Pull request merged**